HW0/P4/P4.py

```diff
@@ -65,12 +101,48 @@
 
     # B = ?
     # C = ?
+    B = np.matrix(np.zeros([6,6]))
+    np.fill_diagonal(B,[bx,by,bz,bvx,bvy,bvz])
+
+    C = np.matrix(np.zeros([3,6]))
+    np.fill_diagonal(C,[rx,ry,rz])
+
 
     # Initial conditions for s0 and Sigma0
+    sigma_0 = 0.01 * np.identity(6)
+
     # Compute the rest of sk using Eqs (2), (3), (4), and (5)
+    def predictS(s_k):
+    	return A*(s_k) + a
+
+    def predictSig(sigma):
+    	return inv(A*sigma*A.T + B*B.T)
+
+    def updateSig(sigma):
+    	return inv(predictSig(sigma) + (C.T)*C)
+
+    def updateS(sigma, s_k, m):
+        return updateSig(sigma)*(predictSig(sigma)*predictS(s_k) + (C.T)*(m))
+
+    s_k2 = np.matrix(np.zeros([6,len(s_measured)+1]))
+    s_k2[:,0] = s_0.T
+
+    s_measured = np.matrix(s_measured)
+
+    for idx, i in enumerate(s_measured):	
+    	if idx == 0:
+        	sigma = sigma_0
+    	else:
+        	sigma = updateSig(sigma)
+    	s_k2[:,idx+1] = updateS(sigma, s_k[:,idx], i.T)
+
+    x_coords = np.array(s_k2)[0].T
+    y_coords = np.array(s_k2)[1].T
+    z_coords = np.array(s_k2)[2].T
+
```

Comment:
Another option would be to multiply the s_measure matrix by the s_stretching matrix instead of appending results of matrix * vector multiplications because python for loops are slower than numpy matrix operations (which are actually done in C)